src/handles/handle.rs

```diff
@@ -0,0 +1,330 @@
+        Pcr24 = ESYS_TR_PCR24,
+        Pcr25 = ESYS_TR_PCR25,
+        Pcr26 = ESYS_TR_PCR26,
+        Pcr27 = ESYS_TR_PCR27,
+        Pcr28 = ESYS_TR_PCR28,
+        Pcr29 = ESYS_TR_PCR29,
+        Pcr30 = ESYS_TR_PCR30,
+        Pcr31 = ESYS_TR_PCR31,
+    }
+
+    impl_constant_handle_conversion_basic!(PcrHandle);
+    impl_constant_handle_conversion!(PcrHandle, ObjectHandle);
+}
+
+/// TPM Constant handle module
+pub mod tpm_constant {
+    use super::object::ObjectHandle;
+    use crate::{
+        tss2_esys::{
+            ESYS_TR, ESYS_TR_RH_ENDORSEMENT, ESYS_TR_RH_LOCKOUT, ESYS_TR_RH_NULL, ESYS_TR_RH_OWNER,
+            ESYS_TR_RH_PLATFORM, ESYS_TR_RH_PLATFORM_NV,
+        },
+        Error, Result, WrapperErrorKind,
+    };
+    use log::error;
+    use num_derive::{FromPrimitive, ToPrimitive};
+    use num_traits::{FromPrimitive, ToPrimitive};
+    use std::convert::{From, TryFrom};
+    /// TpmConstant Handle
+    ///
+    /// Represents an esys handle for
+    /// tpm constant resource handles.
+    #[derive(FromPrimitive, ToPrimitive, Debug, Copy, Clone, PartialEq, Eq)]
+    #[repr(u32)]
+    pub enum TpmConstantHandle {
+        Owner = ESYS_TR_RH_OWNER,
+        Null = ESYS_TR_RH_NULL,
+        Lockout = ESYS_TR_RH_LOCKOUT,
+        Endorsement = ESYS_TR_RH_ENDORSEMENT,
+        Platform = ESYS_TR_RH_PLATFORM,
+        PlatformNv = ESYS_TR_RH_PLATFORM_NV,
+    }
+
+    impl_constant_handle_conversion_basic!(TpmConstantHandle);
+    impl_constant_handle_conversion!(TpmConstantHandle, ObjectHandle);
+}
+
+///
+/// Macro for implmeneting conversion between handles.
+///
+macro_rules! impl_handle_conversion {
+    ($handle_type_self:ident, $handle_type_other:ident) => {
+        impl From<$handle_type_self> for $handle_type_other {
+            fn from(handle_self: $handle_type_self) -> $handle_type_other {
+                $handle_type_other::from(ESYS_TR::from(handle_self.value))
+            }
+        }
+
+        impl From<$handle_type_other> for $handle_type_self {
+            fn from(handle_other: $handle_type_other) -> $handle_type_self {
+                $handle_type_self {
+                    value: handle_other.value(),
+                }
+            }
+        }
+    };
+}
+
+/// Auth handle module
+pub mod auth {
+    use super::object::ObjectHandle;
+    impl_basic_handle!(
+        /// Auth handle
+        ///
+        /// Represents an esys handle to resources
+        /// that can be used for authentication.
+        AuthHandle
+    );
+    impl_handle_conversion!(AuthHandle, ObjectHandle);
+}
+
+/// NV Index handle module
+pub mod nv_index {
+    use super::auth::AuthHandle;
+    use super::object::ObjectHandle;
+    impl_basic_handle!(
+        /// NV Index Handle
+        ///
+        /// Represents an esys resource handle
+        /// for a nv index.
+        NvIndexHandle
+    );
+    impl_handle_conversion!(NvIndexHandle, ObjectHandle);
+    impl_handle_conversion!(NvIndexHandle, AuthHandle);
```

Comment:
Is there a place where these equivalences are listed/mentioned?

Reply:
Not really. Because they are not a real thing in terms of the specification. The specifications I think mentions that the nv operations needs an esys handle to the nv index but for esys everything is ESYS_TR. So that is taken from how I have had to use it. If you look at the tests. There you can see how those conversions are used.

Reply:
They are mainly there to make sure that a key handle is not sent into an API that requires an nv index handle without that 'conversion' explicitly being stated.